File: src/athena/research/supervisor/prepare.py
"""Narrow one-Agent PREPARE phase execution."""

import asyncio
import json
import logging
import subprocess
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from pathlib import Path

from pydantic import BaseModel, ConfigDict, Field

from athena.core.agent.agent_runtime import AgentRuntime
from athena.core.contracts import ArtifactRef, ArtifactStore, CommitHash
from athena.core.tool_types import EmitEvent
from athena.core.workspace import GitWorkBranch, GitWorkspace, GitWorkspaceError
from athena.execution.runtime import ExecutionContext, ExecutionRuntime
from athena.research.evaluation import TrustedEvaluator
from athena.research.evaluation.spec import read_eval_handoff
from athena.research.supervisor.events import wait_run_events
from athena.research.supervisor.experiment import PlanRunner, load_agent_result
from athena.research.supervisor.plans import PlanDecision, PlanInput, PlanState
from athena.research.supervisor.prompt_context import handoff_block

PREPARE_AGENT_ID = "prepare"
PREPARE_PLAN_ID = "prepare"
_REAP_TIMEOUT_SECONDS = 5.0
_FEEDBACK_LIMIT = 1000
logger = logging.getLogger(__name__)
_BaselineGuard = Callable[[], Awaitable[None]]


class PrepareResult(BaseModel):
    """Trusted baseline data returned to the single-writer Supervisor."""

    model_config = ConfigDict(extra="forbid", strict=True)

    evaluator_ref: ArtifactRef
    metric: float = Field(allow_inf_nan=False)
    commit: CommitHash
    predictions_ref: ArtifactRef
    evidence_ref: ArtifactRef
    report_ref: ArtifactRef


class _BaselineGuardAbort(BaseException):
    """Carry a typed guard failure through PlanRunner's evaluator mapping."""

    def __init__(self, error: Exception) -> None:
        super().__init__(str(error))
        self.error = error


class _GuardedEvaluator:
    """Check baseline evidence at the last boundary before trusted scoring."""

    __slots__ = ("_assert_baseline", "_evaluator")

    def __init__(
        self, evaluator: TrustedEvaluator, assert_baseline: _BaselineGuard
    ) -> None:
        self._evaluator = evaluator
        self._assert_baseline = assert_baseline

    async def score(self, **kwargs):
        """Check authority evidence at the final trusted-scoring boundary."""
        try:
            await self._assert_baseline()
        except Exception as error:
            raise _BaselineGuardAbort(error) from error
        return await self._evaluator.score(**kwargs)


@dataclass(frozen=True, slots=True)
class _PrepareDependencies:
    agents: AgentRuntime
    evaluator: TrustedEvaluator
    git: GitWorkspace
    workspace: GitWorkBranch
    execution: ExecutionRuntime
    store: ArtifactStore
    publish: EmitEvent | None


@dataclass(frozen=True, slots=True)
class _PrepareRequest:
    evaluator_ref: ArtifactRef
    tree_ref: ArtifactRef
    task: str
    max_turns: int
    assert_baseline: _BaselineGuard
    predict_features: Path | None = None


async def _reap_agent(agents: AgentRuntime, agent_id: str) -> None:
    """Release the one-shot PREPARE agent within a short cleanup budget."""
    task = asyncio.create_task(agents.reap(agent_id))
    done, _ = await asyncio.wait({task}, timeout=_REAP_TIMEOUT_SECONDS)
    if task not in done:
        task.add_done_callback(_consume_reap_result)
        task.cancel()
        logger.warning("timed out reaping PREPARE Agent %s", agent_id)
        return
    if task.cancelled():
        logger.warning("PREPARE Agent reap was cancelled for %s", agent_id)
        return
    error = task.exception()
    if error is not None:
        logger.warning(
            "failed to reap PREPARE Agent %s",
            agent_id,
            exc_info=(type(error), error, error.__traceback__),
        )


def _consume_reap_result(task: "asyncio.Task[None]") -> None:
    if not task.cancelled():
        task.exception()


def _feedback_from_exception(error: BaseException | str) -> str:
    return " ".join(str(error).split())[:_FEEDBACK_LIMIT]


async def _decision_from_summary(summary, store: ArtifactStore) -> PlanDecision:
    decision = await load_agent_result(summary, store, PlanDecision)
    if decision is None:
        raise RuntimeError(summary.error or "prepare Agent run failed")
    return decision


async def _create_prepare_run(
    deps: _PrepareDependencies,
    request: _PrepareRequest,
    context_ref: ArtifactRef,
) -> str:
    handoff = await read_eval_handoff(deps.store, request.evaluator_ref)
    content = request.task + handoff_block(handoff)
    agent_id, run_id = await deps.agents.create_root(
        PREPARE_AGENT_ID,
        {"content": content, "context_refs": [context_ref]},
        agent_id=PREPARE_AGENT_ID,
        name=PREPARE_PLAN_ID,
    )
    if agent_id != PREPARE_AGENT_ID:
        raise RuntimeError(f"prepare Agent id must be {PREPARE_AGENT_ID}")
    return run_id


def _prepare_plan_state(
    context_ref: ArtifactRef, turn: int, max_turns: int
) -> PlanState:
    return PlanState(
        kind="PREPARE",
        context_ref=context_ref,
        turns_used=turn,
        turn_limit=max_turns,
    )


def _prepare_plan_input(request: _PrepareRequest) -> PlanInput:
    return PlanInput(evaluator_ref=request.evaluator_ref, tree_ref=request.tree_ref)


def _assert_scored_outcome(
    outcome,
    decision: str,
) -> tuple[float, CommitHash, ArtifactRef, ArtifactRef, ArtifactRef]:
    if outcome.kind == "evaluator_infrastructure_failed":
        raise RuntimeError(outcome.error or "evaluator unavailable")
    if outcome.kind != "scored":
        raise ValueError(outcome.error or f"PREPARE validation failed: {outcome.kind}")
    if outcome.commit is None:
        raise ValueError("trusted commit is missing")
    if outcome.predictions_ref is None:
        raise ValueError("trusted predictions are missing")
    if outcome.evidence_ref is None:
        raise ValueError("trusted evidence is missing")
    if outcome.report_ref is None:
        raise ValueError("trusted report is missing")
    if outcome.metric is None:
        raise ValueError("trusted metric is missing")
    if decision != "submit":
        raise ValueError(
            "baseline validated successfully "
            f"(metric {outcome.metric:.4f}) but the decision was "
            f"{decision!r}. Return submit to advance to SEARCH; "
            "continue means keep repairing this baseline, not move on."
        )
    return (
        outcome.metric,
        outcome.commit,
        outcome.predictions_ref,
        outcome.evidence_ref,
        outcome.report_ref,
    )


async def run_prepare_plan(
    *,
    agents: AgentRuntime,
    evaluator: TrustedEvaluator,
    git: GitWorkspace,
    workspace: GitWorkBranch,
    execution: ExecutionRuntime,
    store: ArtifactStore,
    evaluator_ref: ArtifactRef,
    tree_ref: ArtifactRef,
    task: str,
    max_turns: int,
    assert_baseline: _BaselineGuard,
    publish: EmitEvent | None = None,
    predict_features: Path | None = None,
) -> PrepareResult:
    """Run and repair one stable PREPARE Agent until a trusted baseline exists."""
    if max_turns < 1:
        raise ValueError("max_turns must be at least 1")

    deps = _PrepareDependencies(
        agents=agents,
        evaluator=evaluator,
        git=git,
        workspace=workspace,
        execution=execution,
        store=store,
        publish=publish,
    )
    request = _PrepareRequest(
        evaluator_ref=evaluator_ref,
        tree_ref=tree_ref,
        task=task,
        max_turns=max_turns,
        assert_baseline=assert_baseline,
        predict_features=predict_features,
    )

    deps.execution.ensure_environment()
    root = Path(workspace.path).resolve()
    root.mkdir(parents=True, exist_ok=True)

    context_ref = await store.put_text(
        json.dumps(
            {"plan_id": PREPARE_PLAN_ID, "task": task, "tree_ref": tree_ref},
            ensure_ascii=False,
        )
    )
    run_id = await _create_prepare_run(deps, request, context_ref)

    runner = PlanRunner(
        execution=execution,
        store=store,
        evaluator=_GuardedEvaluator(evaluator, assert_baseline),
        workspace=git,
        branch=workspace,
        context=ExecutionContext(
            project_root=execution.project_root,
            workspace_root=root,
            environment_root=execution.environment_root,
            experiment_id=PREPARE_PLAN_ID,
            predict_features=predict_features,
        ),
    )

    feedback: str | None = None
    try:
        for turn in range(max_turns):
            if turn:
                run_id = await agents.followup(
                    PREPARE_AGENT_ID,
                    {"content": feedback, "context_refs": []},
                )

            summary = await wait_run_events(agents, run_id, publish)
            await assert_baseline()
            try:
                decision = await _decision_from_summary(summary, store)
            except (OSError, RuntimeError, ValueError) as error:
                feedback = (
                    "previous Agent turn did not produce a valid decision: "
                    f"{_feedback_from_exception(error)}"
                )
                continue
            if decision.decision == "abandon":
                raise RuntimeError(f"prepare Agent abandoned Plan: {decision.reason}")

            try:
                outcome = await runner.run_turn(
                    PREPARE_PLAN_ID,
                    _prepare_plan_state(context_ref, turn, max_turns),
                    _prepare_plan_input(request),
                    emit=publish,
                )
                metric, commit, predictions_ref, evidence_ref, report_ref = (
                    _assert_scored_outcome(outcome, decision.decision)
                )
                return PrepareResult(
                    evaluator_ref=evaluator_ref,
                    metric=metric,
                    commit=commit,
                    predictions_ref=predictions_ref,
                    evidence_ref=evidence_ref,
                    report_ref=report_ref,
                )
            except _BaselineGuardAbort as error:
                if isinstance(error.error, RuntimeError):
                    raise error.error
                raise RuntimeError("baseline evidence guard failed") from error.error
            except (
                OSError,
                ValueError,
                subprocess.SubprocessError,
                GitWorkspaceError,
            ) as error:
                feedback = _feedback_from_exception(error)
                continue
    finally:
        await _reap_agent(agents, PREPARE_AGENT_ID)

    raise RuntimeError("prepare turn budget exhausted without a trusted baseline")


__all__ = [
    "PREPARE_AGENT_ID",
    "PREPARE_PLAN_ID",
    "PrepareResult",
    "run_prepare_plan",
]

File: athena_ts/packages/athena-research/src/supervisor/prepare.ts
/**
 * PREPARE 阶段合同（移植 ``research/supervisor/prepare.py`` 的 PrepareResult 数据模型
 * 与 ``_freeze_evaluator`` 确定性冻结逻辑）。
 */

import { existsSync, readFileSync, statSync } from "node:fs"
import { dirname, join, relative } from "node:path"
import { z } from "zod"
import {
  ArtifactRef,
  CommitHash,
  FiniteFloat,
  type ArtifactStore,
  type GitWorkBranch,
  type GitWorkspace,
} from "@athena/core"

import { DataScriptBundleSchema, PlanInputSchema, PlanStateSchema, type PlanDecision } from "../contracts.js"
import type { DataScriptRunner } from "../script_runner.js"
import type { Scorer } from "../evaluation.js"
import type { ExecutionRuntime } from "../execution.js"
import { PlanRunner, hasAnyFile } from "./experiment.js"
import { saveResearchState, type ResearchState } from "./state.js"

export const PREPARE_PLAN_ID = "prepare"

/** Trusted baseline data returned to the single-writer Supervisor。 */
export const PrepareResultSchema = z.strictObject({
  evaluator_ref: ArtifactRef,
  metric: FiniteFloat,
  commit: CommitHash,
  predictions_ref: ArtifactRef,
  evidence_ref: ArtifactRef,
  report_ref: ArtifactRef,
})
export type PrepareResult = z.infer<typeof PrepareResultSchema>

/**
 * 建立 PREPARE 的稳定实验目录（对齐 Python PhaseRunner）：
 * 初始化 `.athena/repo` → 创建固定 `workspaces/eda` worktree → 立即持久化 eda_dir。
 */
export async function createPrepareWorkspace(opts: {
  git: GitWorkspace
  projectRoot: string
  state: ResearchState
  statePath: string
}): Promise<GitWorkBranch> {
  const baseCommit = await opts.git.init(undefined, ".gitignore", ".venv/\n")
  const workspace = await opts.git.create(baseCommit, "athena/prepare", { name: "eda" })
  opts.state.eda_dir = relative(opts.projectRoot, workspace.path) || "."
  saveResearchState(opts.statePath, opts.state)
  return workspace
}

/** 冻结 evaluator 目录（metric.json 的 eval_script）为 bundle，返回 bundle JSON ref。 */
export async function freezeEvaluator(opts: {
  root: string
  scripts: DataScriptRunner
  store: ArtifactStore
}): Promise<ArtifactRef> {
  const { root, scripts, store } = opts
  const specPath = join(root, "metric.json")
  if (!existsSync(specPath)) throw new Error("metric.json is missing")
  let evaluatorRel: string
  try {
    const spec = JSON.parse(readFileSync(specPath, "utf-8")) as { eval_script?: unknown }
    if (typeof spec.eval_script !== "string") throw new Error()
    evaluatorRel = spec.eval_script
  } catch {
    throw new Error("metric.json must declare eval_script")
  }
  const evaluatorPath = join(root, evaluatorRel)
  let evaluatorRoot: string
  let entrypoint: string
  if (existsSync(evaluatorPath) && statSync(evaluatorPath).isDirectory()) {
    evaluatorRoot = evaluatorPath
    entrypoint = "evaluate.py"
    if (!existsSync(join(evaluatorRoot, entrypoint))) {
      throw new Error(`eval_script directory must contain an entrypoint file named evaluate.py: ${JSON.stringify(evaluatorRel)}`)
    }
  } else if (existsSync(evaluatorPath) && statSync(evaluatorPath).isFile()) {
    evaluatorRoot = dirname(evaluatorPath)
    entrypoint = evaluatorRel.split(/[\\/]/).pop()!
  } else {
    throw new Error("eval_script is missing")
  }

  const labelsFile = join(evaluatorRoot, "labels.csv")
  const labelsDir = join(evaluatorRoot, "labels")
  const hasLabels =
    (existsSync(labelsFile) && statSync(labelsFile).size > 0) ||
    (existsSync(labelsDir) && hasAnyFile(labelsDir))
  if (!hasLabels) {
    throw new Error(
      "eval labels are missing: labels.csv or a non-empty labels/ dir must sit next to the eval script"
    )
  }

  const bundle = await scripts.freeze(evaluatorRoot, entrypoint)
  return store.putText(JSON.stringify(DataScriptBundleSchema.parse(bundle)))
}

/** 运行一次 worker agent turn 的抽象（返回 PlanDecision 或 null）。 */
export interface AgentTurn {
  turn(prompt: string): Promise<PlanDecision | null>
}

/** 运行 evaluator agent 直到冻结出一个 evaluator bundle。 */
export async function runEvaluatorPlan(opts: {
  agent: AgentTurn
  scripts: DataScriptRunner
  store: ArtifactStore
  evaluatorDir: string
  task: string
  maxTurns: number
}): Promise<ArtifactRef> {
  if (opts.maxTurns < 1) throw new Error("max_turns must be at least 1")
  let feedback: string | null = null
  for (let turn = 0; turn < opts.maxTurns; turn++) {
    const prompt = feedback ?? opts.task
    const decision = await opts.agent.turn(prompt)
    if (decision === null) {
      feedback = "previous evaluator turn did not produce a valid decision"
      continue
    }
    if (decision.decision === "abandon") {
      throw new Error(`evaluator Agent abandoned Plan: ${decision.reason}`)
    }
    try {
      const evaluatorRef = await freezeEvaluator({
        root: opts.evaluatorDir,
        scripts: opts.scripts,
        store: opts.store,
      })
      if (decision.decision !== "submit") {
        throw new Error(`evaluator frozen successfully but the decision was ${decision.decision}`)
      }
      return evaluatorRef
    } catch (exc) {
      feedback = (exc as Error).message
    }
  }
  throw new Error("evaluator turn budget exhausted without a frozen evaluator")
}

/** 运行 prepare agent 直到产生可信 baseline。 */
export async function runPreparePlan(opts: {
  agent: AgentTurn
  evaluator: Scorer
  git: GitWorkspace
  workspace: GitWorkBranch
  execution: ExecutionRuntime
  store: ArtifactStore
  evaluatorRef: ArtifactRef
  treeRef: ArtifactRef
  task: string
  maxTurns: number
}): Promise<PrepareResult> {
  if (opts.maxTurns < 1) throw new Error("max_turns must be at least 1")
  const contextRef = await opts.store.putText(
    JSON.stringify({ plan_id: PREPARE_PLAN_ID, task: opts.task, tree_ref: opts.treeRef })
  )
  let feedback: string | null = null
  for (let turn = 0; turn < opts.maxTurns; turn++) {
    const prompt = feedback ?? opts.task
    const decision = await opts.agent.turn(prompt)
    if (decision === null) {
      feedback = "previous Agent turn did not produce a valid decision"
      continue
    }
    if (decision.decision === "abandon") {
      throw new Error(`prepare Agent abandoned Plan: ${decision.reason}`)
    }
    try {
      const runner = new PlanRunner(opts.execution, opts.store, opts.evaluator, opts.git, opts.workspace)
      const state = PlanStateSchema.parse({
        kind: "PREPARE",
        context_ref: contextRef,
        turns_used: turn,
        turn_limit: opts.maxTurns,
      })
      const planInput = PlanInputSchema.parse({
        evaluator_ref: opts.evaluatorRef,
        tree_ref: opts.treeRef,
      })
      const outcome = await runner.runTurn(PREPARE_PLAN_ID, state, planInput)
      if (outcome.kind === "evaluator_infrastructure_failed") {
        throw new Error(outcome.error ?? "evaluator unavailable")
      }
      if (outcome.kind !== "scored") {
        throw new Error(outcome.error ?? `PREPARE validation failed: ${outcome.kind}`)
      }
      if (outcome.commit === null || outcome.predictions_ref === null || outcome.evidence_ref === null || outcome.report_ref === null || outcome.metric === null) {
        throw new Error("trusted baseline is missing required outputs")
      }
      if (decision.decision !== "submit") {
        throw new Error(`baseline validated successfully but the decision was ${decision.decision}`)
      }
      return PrepareResultSchema.parse({
        evaluator_ref: opts.evaluatorRef,
        metric: outcome.metric,
        commit: outcome.commit,
        predictions_ref: outcome.predictions_ref,
        evidence_ref: outcome.evidence_ref,
        report_ref: outcome.report_ref,
      })
    } catch (exc) {
      if (exc instanceof Error && (exc.message.includes("evaluator unavailable") || exc.message.includes("trusted baseline is missing"))) {
        throw exc
      }
      feedback = (exc as Error).message
    }
  }
  throw new Error("prepare turn budget exhausted without a trusted baseline")
}
